### PR TITLE
Convert lab/collaborators to JSX

### PR DESCRIPTION
**Pull request by eatyourgreens**
- Rewrite the old coffeescript collaborators page as JSX.
- Replace `createReactClass` with functional components.
- Replace `PromiseRenderer` with data-fetching hooks.

Adding project team members, removing them and changing their roles should all work on this branch. The coffeescript code for the event handlers was hard to follow, but I think I've reproduced it. The code for Talk roles was very hard to follow, so I think that needs double-checking.

Staging branch URL: https://pr-6233.pfe-preview.zooniverse.org

Towards #3216.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `npm ci` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on GitHub Actions?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Affects: Project Builder -> Collaborators page

This PR updates the code for the Collaborators page, moving from CoffeeScript to use JavaScript. No behaviour/functionality changes expected.

_(Note: the page labels all members as "collaborators", though a "collaborator" is also a specific role. For this review, I'll refer to the collection of users who are collaborators(role)/experts/researchers/etc as "members" of the project.)_

Collaborators page functionality checklist:
- The page [(example)](https://local.zooniverse.org:3735/lab/1986/collaborators) should be accessible to (viewable, editable by) (1) the project owner, (2) collaborators, and (3) Zooniverse admins. 
- The page should indicate who's the project owner.
- The page should (1) list all users who are members of the project, and (2) their roles on the project.
  - Members can be deleted by clicking on the 'x' button next to their logins IDs.
  - Members can have their roles changed by clicking on the appropriate checkboxes.
- The page should allow new members to be added.
  - This is done by (1) searching for usernames, (2) selecting roles, and (3) clicking the "Add User Role" button
  - Note: if a user is already a member, then attempting to add the user again (even with different roles) will result in the error _"Error: Validation failed: User group . Roles have already been set for this user or group"_
- The `museum` and `translator` roles should appear in (1) the members list and (2) add new member panel, IF the appropriate experimental tools are enabled.

### Testing

Tested with macOS + Chrome on localhost staging.

Testing:
- Login as User A (e.g. `darkeshard`) and create Project X [(example)](https://local.zooniverse.org:3735/lab/1986)
- Go to Project X's collaborator page [(example)](https://local.zooniverse.org:3735/lab/1986/collaborators)
  - Check that page says "you are the owner of this project"
  - Check that no members are listed.
  - Add User B (e.g. `zootester1`) as a member with "collaborator" role.
  - Add User C (e.g. `zootester2`) as a member with "researcher" role.
  - Reload the page. Check that User B and C are now listed as members
  - Modify User C's roles, e.g. adding "moderator". Reload the page to confirm.
    - Advanced: keep removing and adding random roles to User C. (Don't rapidly click checkboxes; wait a second to ensure each change is sent properly to the server)
      - ❓ Expectation: roles should be added/removed with no issues. Actual: roles are added/removed as expected, but an error randomly pops up when adding a role to User C, removing that role, _and then adding it again._ See **Issue 1.**
      - Try to add User B or C again.
    - ❗ Expectation: receive an error message. Actual: error message is obscured. See **Issue 2.**
  - Remove User C from the list of members. Reload to confirm.
- In a separate window, login as User B and visit Project X's Collaborator page.
  - Check that page says "User A is the owner of this project"
  - Check that User B is listed as a member.
  - Add User C as a member with "researcher" role. Reload the page to confirm.
  - Repeat the other testing steps above.

Basic testing looks good, but there are some low-key issues worth noting.

**Issue 1:** Adding a role to a member that _previously had and then lost_ that role results in an error.

- The role is STILL added properly. This can be confirmed by reloading the page.
- The error message on the page is garbled (see Issue 2), but the dev console shows the familiar _"Validation failed: User Role already exists for this user and section"_ warning/error.

_Video below: for context, I'm logged in as darkeshard; and zootester1 previously had the "researcher" role, and then I removed it. In the video, I click on zootester1's "researcher" checkbox to add the role. The role is added, but an error message appears. The dev console shows a warning._
https://user-images.githubusercontent.com/13952701/199856530-a179922e-7fd3-4c07-b7ec-382130100622.mov

I'm noting this as an issue ONLY because I've been unable to _reliably_ replicate it on `www.zooniverse.org`, which makes me think the problem is something introduced in this PR. (Maybe?? Note the suspiciously italicised _reliably_ in that sentence.) That said, the error is clearly coming from the API, _yet adding the role still works,_ so this is less of a "it doesn't work as advertised" issue, and more of a "the UI can be confusing/misleading in a very specific scenario" issue.

This issue is not a blocker for this PR, though I would like to dig more into it.

**Issue 2:** the error messages aren't being displayed properly.

- Errors are being displayed as `[object Object]` instead of the actual error message.
- NOTE: the issue was noticed on the CollaboratorCreator component, but may also apply to the parent Collaborator (listing) component.

_Screenshot below: on www.zooniverse.org, when attempting to add a user who's already a member, the error message appears correctly._
Screenshot 2022-11-03 at 23 38 36

_Screenshot below: on localhost:3735 on the branch, the same action leads to a less useful error message._
Screenshot 2022-11-03 at 23 38 56

This issue isn't preventing me from approving this PR, though it'd be nice to get this fixed before a merge. I think it should be a pretty straightforward debug of the error data structure being caught during the new fetches.

SIDE NOTE: I ran into some issues testing the more advanced items, as by 11pm UK time, I'm getting extremely slow responses from Panoptes Staging. Typing `zootester1` letter-by-letter causes the "Add User" field to "hang" (response from Panoptes Staging times out), possibly due to me basically "spamming" the user search API endpoint by typing so much. Not sure what to make of this.

BONUS QUIRK:
- Not related to this PR, but interesting to note: I tried something goofy where I'm logged in as User B (the project collaborator), and then tried to _remove myself as a member of the project,_ or _remove the collaborator role from myself._
- This resulted in shenanigans, as expected. I think removing myself as a member resulted in a weird (but technically non-harmful) zombie where a non-member can still view (but not edit) the project.
- Removing the collaborator role on the other hand resulted in a bizarre error message I haven't been able to reliably replicate, due to the Panoptes Staging issues.

### Status

I'm giving this a LGTM since the basic tests pass, but it'd be good if those 2 issues can be resolved before merging.

**Comment by eatyourgreens** [comment]
The error message bug was already fixed in #6237.

File: app/pages/lab/collaborators.jsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    .catch(setError)
+    .then(() => {
+      project.uncacheLink('project_roles');
+      const savingIndex = saving.indexOf(projectRoleSet.id);
+      setSaving(saving.splice(savingIndex, 1));
+      setProjectRoleSets(projectRoleSets.filter(roleSet => roleSet.id !== projectRoleSet.id));
+    });
+  }
+
+  function addCollaborator() {
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply app/pages/lab/collaborators.jsx:122]:
```suggestion
  function refreshProjectRoles() {
```

File: app/pages/lab/collaborators.jsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        }
+      </div>
+
+      <hr />
+
+      <div className="form-label">Add another</div>
+      <CollaboratorCreator
+        possibleRoles={possibleRoles}
+        project={project}
+        onAdd={addCollaborator}
```
Comment (eatyourgreens) [inline app/pages/lab/collaborators.jsx:173]:
```suggestion
        onAdd={refreshProjectRoles}
```

**Comment by eatyourgreens** [comment]
I've been able to trigger the 'Validation failed' error by toggling roles on and off on staging here, which is still using the coffeescript version of the page.
https://master.pfe-preview.zooniverse.org/lab/16957/collaborators

File: app/pages/lab/collaborators.jsx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        .create({
+          user_id: parseInt(projectRoleSet.links.owner.id),
+          section: projectSection(project),
+          name: possibleRoles[role]
+        })
+        .save();
+      }
+    } else {
+      projectRoleSet.roles.splice(index, 1);
+      const filteredRoles = projectRoleSet.talk_roles.filter(talkRole => talkRole === possibleRoles[role]);
```
Comment (eatyourgreens) [inline app/pages/lab/collaborators.jsx:93]: This uses `talkRole.name` in the original coffeescript.
```suggestion
      const filteredRoles = projectRoleSet.talk_roles.filter(talkRole => talkRole.name === possibleRoles[role]);
```

**Comment by eatyourgreens** [comment]
The API errors are coming back from the Talk API, so I think something is wrong in the code that handles Talk roles.

**Comment by eatyourgreens** [comment]
@shaunanoordin I read back through the new code, comparing with the coffeescript version, and found a few errors in the `toggleRole` handler.
